### DISCUSSION TITLE
Possible to purge historical data?

### DISCUSSION THREAD

**Discussion by benr77** [zitadel/zitadel#7617] [Q&A] [answered]
I have had an issue with a machine user repeatedly trying to create a Zitadel user via an API call, which was badly configured and constantly failing. 

This occurred over a period of days and a huge amount of data seems to have been stored in the Zitadel database.

I don't really want to carry this overhead for all eternity as it's all redundant data.

Is it possible to purge Zitadel data for things like this? Either per user or event type or perhaps date range?

**Comment by hifabienne** [comment]
Hi @benr77 
Unfortunately it is not easily possible to purge data.
Due to our architecture, where our goal is to store everything that happens as an event to make it possible to see the whole audit log, you will always have the events even if you perform a delete request.

One possiblity would be to just perform delete requests on your resources. You will not see them listed anymore, but still have the events store.

**Comment by benr77** [comment]
The create user requests were failing, so in my case I don't have any users to delete/hide. It's just concerning the data stored for all these pointless requests.

We are still in the prototyping stage with our platform - we have currently only 10 users, with possibly 100 test users that have been deleted. Already our database is storing 1GB of data, and before this particular problem it was already 750MB.

What is the strategy for managing database size when we have thousands of users and potentially tens of thousands of events per day? Surely the DB will just get enormous. 

We have no requirement to store historical data since day 0, and do not really want to be paying a cloud provider for huge database storage costs for data we are not interested in.

Any info would be appreciated.

**Reply by hifabienne** [reply]
Hi @benr77 
So first to say, one of our unique selling point is that we store all the historical data since day 0, as we want to provide our customers with the whole audit trail. So this will not change in the future. 

But I am not sure where the problem is in your case. When a user can't be created and there is an error, ZITADEL does not store anything to the database. So this is probably not the reason why you use so much storage.

One thing I can think of it, that we do store a token each time a user authenticates. Might this be the problem?

**Reply by adlerhurst** [reply]
can you please check the space used per table and share the output of the following command?

and which version of zitadel do you use?

```sql
select
  concat(quote_ident(table_schema), '.', quote_ident(table_name)) "table",
  pg_size_pretty(pg_total_relation_size(concat(quote_ident(table_schema), '.', quote_ident(table_name)))),
  pg_total_relation_size(concat(quote_ident(table_schema), '.', quote_ident(table_name)))
from information_schema.tables
where table_schema = any(array['projections', 'eventstore', 'auth', 'logstore', 'adminapi', 'system'])
order by 3 desc;
```

**Comment by benr77** [comment]
It's the very latest Zitadel 2.48.1. Here is the output.

table | pg_size_pretty | pg_total_relation_size
-- | -- | --
eventstore.events2 | 94 MB | 98369536
auth.tokens | 19 MB | 20103168
auth.auth_requests | 360 kB | 368640
projections.current_states | 152 kB | 155648
projections.keys4_private | 128 kB | 131072
system.assets | 128 kB | 131072
projections.users10 | 96 kB | 98304
projections.custom_texts2 | 88 kB | 90112
projections.login_names3_users | 80 kB | 81920
projections.message_texts2 | 80 kB | 81920
eventstore.unique_constraints | 72 kB | 73728
projections.keys4_public | 64 kB | 65536
projections.login_names3_domains | 64 kB | 65536
projections.mail_templates2 | 64 kB | 65536
projections.project_members4 | 64 kB | 65536
projections.user_grants4 | 64 kB | 65536
projections.orgs1 | 64 kB | 65536
projections.user_grants5 | 64 kB | 65536
projections.org_members4 | 64 kB | 65536
projections.instance_members4 | 64 kB | 65536
projections.apps6_saml_configs | 64 kB | 65536
projections.user_auth_methods4 | 64 kB | 65536
projections.apps6 | 48 kB | 49152
projections.domain_policies2 | 48 kB | 49152
projections.privacy_policies3 | 48 kB | 49152
projections.instance_domains | 48 kB | 49152
projections.password_complexity_policies2 | 48 kB | 49152
projections.password_age_policies2 | 48 kB | 49152
projections.lockout_policies2 | 48 kB | 49152
auth.users2 | 48 kB | 49152
projections.label_policies3 | 48 kB | 49152
projections.users10_notifications | 48 kB | 49152
auth.user_sessions | 48 kB | 49152
projections.user_metadata5 | 48 kB | 49152
projections.failed_events2 | 48 kB | 49152
projections.apps6_oidc_configs | 48 kB | 49152
projections.login_policies5 | 48 kB | 49152
projections.org_domains2 | 48 kB | 49152
adminapi.styling2 | 48 kB | 49152
projections.login_names3_policies | 48 kB | 49152
projections.apps6_api_configs | 48 kB | 49152
projections.projects4 | 48 kB | 49152
projections.personal_access_tokens3 | 40 kB | 40960
system.encryption_keys | 32 kB | 32768
projections.authn_keys2 | 32 kB | 32768
projections.users10_humans | 32 kB | 32768
projections.users10_machines | 32 kB | 32768
projections.actions3 | 32 kB | 32768
projections.idp_user_links3 | 32 kB | 32768
projections.idps3 | 32 kB | 32768
projections.instance_features2 | 32 kB | 32768
projections.instances | 32 kB | 32768
projections.keys4 | 32 kB | 32768
projections.keys4_certificate | 32 kB | 32768
projections.idp_login_policy_links5 | 32 kB | 32768
projections.locks | 32 kB | 32768
projections.milestones | 32 kB | 32768
projections.notification_policies | 32 kB | 32768
projections.idp_templates5 | 32 kB | 32768
projections.oidc_settings2 | 32 kB | 32768
projections.project_grant_members4 | 32 kB | 32768
projections.project_grants4 | 32 kB | 32768
projections.project_roles4 | 32 kB | 32768
projections.restrictions2 | 32 kB | 32768
projections.secret_generators2 | 32 kB | 32768
projections.security_policies2 | 32 kB | 32768
projections.smtp_configs1 | 32 kB | 32768
adminapi.current_sequences | 24 kB | 24576
projections.current_sequences | 24 kB | 24576
projections.device_auth_requests | 24 kB | 24576
projections.sessions8 | 24 kB | 24576
auth.failed_events | 24 kB | 24576
auth.current_sequences | 24 kB | 24576
adminapi.failed_events | 24 kB | 24576
auth.user_external_idps2 | 24 kB | 24576
projections.org_metadata2 | 24 kB | 24576
auth.refresh_tokens | 24 kB | 24576
auth.org_project_mapping2 | 24 kB | 24576
projections.failed_events | 24 kB | 24576
auth.idp_providers2 | 24 kB | 24576
auth.idp_configs2 | 24 kB | 24576
projections.quotas_notifications | 16 kB | 16384
projections.auth_requests | 16 kB | 16384
projections.executions | 16 kB | 16384
projections.flow_triggers3 | 16 kB | 16384
projections.idp_templates5_apple | 16 kB | 16384
projections.idp_templates5_azure | 16 kB | 16384
projections.idp_templates5_github | 16 kB | 16384
projections.idp_templates5_github_enterprise | 16 kB | 16384
projections.idp_templates5_gitlab | 16 kB | 16384
projections.idp_templates5_gitlab_self_hosted | 16 kB | 16384
projections.idp_templates5_google | 16 kB | 16384
projections.idp_templates5_jwt | 16 kB | 16384
projections.idp_templates5_ldap2 | 16 kB | 16384
projections.idp_templates5_oauth2 | 16 kB | 16384
projections.idp_templates5_oidc | 16 kB | 16384
projections.idp_templates5_saml | 16 kB | 16384
projections.idps3_jwt_config | 16 kB | 16384
projections.idps3_oidc_config | 16 kB | 16384
projections.limits | 16 kB | 16384
projections.notification_providers | 16 kB | 16384
projections.quotas | 16 kB | 16384
projections.quotas_periods | 16 kB | 16384
projections.sms_configs2 | 16 kB | 16384
projections.sms_configs2_twilio | 16 kB | 16384
projections.system_features | 16 kB | 16384
projections.targets | 16 kB | 16384
adminapi.locks | 16 kB | 16384
adminapi.styling | 16 kB | 16384
auth.idp_configs | 16 kB | 16384
auth.idp_providers | 16 kB | 16384
auth.locks | 16 kB | 16384
auth.org_project_mapping | 16 kB | 16384
auth.user_external_idps | 16 kB | 16384
auth.users | 16 kB | 16384
logstore.execution | 16 kB | 16384
logstore.access | 16 kB | 16384
projections.login_names3 | 0 bytes | 0

What I don't understand is that this looks like it's using ~100MB in total, which is pretty reasonable. However, on the Digital Ocean Managed Database the disk usage is reporting 9.4% of 10GB, which is ~1GB. The discrepancy is a whole order of magnitude...

**Comment by benr77** [comment]
Further to this, I've been reading that Digital Ocean Managed Databases disk usage report is also including indexes, log data etc, which would explain some of the discrepancy in disk space used, but surely not 887MB of overhead for 113MB of data?

**Reply by fforootd** [reply]
This is a rather odd thing, but I think we can get down to the cause here 😁 

Usually our database is quiet efficient to store things since a lot of the information is redundant (i.e. the events) an can be crompressed with ease. Actually only the payload and the ID is unique data.

**Reply by fforootd** [reply]
Just a thought occured... can it be that point in time recovery uses the space or the WAL files in general?

**Reply by benr77** [reply]
I have [posted a question at Digital Ocean](https://www.digitalocean.com/community/questions/managed-postgres-instance-reporting-10x-disk-space-used-vs-data-stored) regarding this

**Reply by fforootd** [reply]
Thanks, lets have a look on the response ;-)

**Comment by benr77** [comment, marked as answer]
OK, so based on what I've discovered, here is an answer to the original question.

- **Zitadel cannot purge anything from the event store**, due to this being a specific part of the architecture. This is not a problem because....
- the disk space used by the Zitadel database is not really that large, and...
- Postgres, and presumably other DBs, are storing a lot of data on disk outside of the raw table data (logs, indexes etc) and this can cause your disk usage to vary wildly from your table space usage.
- automatic management of log files by Postgres can for example *reduce* your disk space usage once a period of heavy activity has subsided.

Thanks for all the input from everyone.

**Reply by fforootd** [reply]
Thanks for sharing this, I am certain this helps others too.
